ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -87,28 +89,40 @@ private[platform] final class EventsBuffer[O: Ordering, E](
     * @param endInclusive The end inclusive bound of the requested range.
     * @return The series of events as an ordered vector satisfying the input bounds.
     */
-  def slice(startExclusive: O, endInclusive: O): BufferSlice[(O, E)] =
+  def slice[FILTER_RESULT](
+      startExclusive: Offset,
+      endInclusive: Offset,
+      filter: ENTRY => Option[FILTER_RESULT],
+  ): BufferSlice[(Offset, FILTER_RESULT)] =
     Timed.value(
       sliceTimer, {
         val bufferSnapshot = _bufferStateRef
+        val vectorSnapshot = bufferSnapshot.vector
+
         if (bufferSnapshot.rangeEnd.exists(_ < endInclusive)) {
-          throw RequestOffBufferBounds(bufferSnapshot.vector.last._1, endInclusive)
-        } else if (bufferSnapshot.vector.isEmpty)
-          Empty
-        else {
-          val Seq(bufferStartInclusiveIdx, bufferEndExclusiveIdx) =
-            Seq(startExclusive, endInclusive)
-              .map(bufferSnapshot.vector.searchBy(_, _._1))
-              .map {
-                case InsertionPoint(insertionPoint) => insertionPoint
-                case Found(foundIndex) => foundIndex + 1
-              }
-
-          val vectorSlice =
-            bufferSnapshot.vector.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
-
-          if (bufferStartInclusiveIdx == 0) Prefix(vectorSlice)
-          else Inclusive(vectorSlice)
+          throw RequestOffBufferBounds(bufferSnapshot.rangeEnd.get, endInclusive)
+        } else {
+          val bufferStartSearchResult = vectorSnapshot.searchBy(startExclusive, _._1)
+          val bufferEndSearchResult = vectorSnapshot.searchBy(endInclusive, _._1)
+
+          val bufferStartInclusiveIdx = indexAfter(bufferStartSearchResult)
+          val bufferEndExclusiveIdx = indexAfter(bufferEndSearchResult)
+
+          val bufferSlice = vectorSnapshot.slice(bufferStartInclusiveIdx, bufferEndExclusiveIdx)
+
+          bufferStartSearchResult match {
+            case InsertionPoint(0) if bufferSlice.isEmpty =>
+              Suffix(endInclusive, Vector.empty, None)
+            case InsertionPoint(0) =>
+              val (filteredSlice, continueFrom) = filterAndChunkSlice(bufferSlice.tail, filter)
```

Comment:
As discussed offline we could have here the chunk calculated from backwards to struck a trade-off between more versatile stream-catch-up and having stream catch-up issues depend on the max buffer size, versus simpler logic but less versatile stream catch up, which depends on the max chink size.

Reply:
Done

Reply:
:raised_hands: